### PR TITLE
T13934

### PR DESCRIPTION
**Pull request by DieKatze**
Hi @gbastien,

Would you mind merging this, if it pass your code review ofc.

Thanks.

DieKatze.

### CODE REVIEW

File: src/imio/helpers/date.py [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+    ############# globals ################
+    ones = ["", "un ","deux ","trois ","quatre ", "cinq ",
+    "six ","sept ","huit ","neuf "]
+    tens = ["dix ","onze ","douze ","treize ", "quatorze ",
+    "quinze ","seize ","dix-sept ","dix-huit ","dix-neuf "]
+    twenties = ["","","vingt ","trente ","quarante ",
+    "cinquante ","soixante ","septante ","quatre-vingt ","nonante "]
+    thousands = ["","mille ","million ", "milliard ", "billiard ",
+    "trilliard ", "quadrilliard ", "quintillard ", "sextilliard", "septilliard ",
+    "octilliard ", "nonilliard ", u"dÃ©cillion ", u"dÃ©cilliard ", u"unodÃ©cillion ", u"unodÃ©cilliard ", u"duodÃ©cillion ", u"duodÃ©cilliard "]
```
Comment (gbastien) [inline src/imio/helpers/date.py]: WTF is this?

File: src/imio/helpers/tests/test_date.py [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+    def test_formatDate(self):
+        self.assertEqual(formatDate("now", month_name=False), DateTime().strftime("%d/%m/%Y"))
+        self.assertEqual(formatDate(DateTime(2,1,1,1,1), month_name=True, long_format=True), u'1 janvier 2002 (01:01)')
+
+    def test_int2word(self):
+        self.assertEqual(int2word(21), "vingt et un")
+        self.assertEqual(int2word(100), "cent")
+        self.assertEqual(int2word(220), "deux cent vingt")
+        self.assertEqual(int2word(1000100), "un million cent")
```
Comment (gbastien) [inline src/imio/helpers/tests/test_date.py]: Could you add a test with one of the "thousands" using an accentuated value?

File: src/imio/helpers/date.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+    if long_format:
+        hour = ' (%H:%M)'
+    else:
+        hour = ''
+    #check if we need to return the month name
+    if month_name:
+        month = date.month()
+        translated_month = translate(msgid=monthsIds[month], domain="plonelocales", context=context, default=monthsIds[month], target_language='fr')
+        #use mOntH here to avoid strftime translating the month...
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Please try to limit length to 80 characters if possible, 120 max anyway.

File: src/imio/helpers/date.py [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+            if q >= 0:
+                n3.append(int(r[:3]))
+            elif q >= -1:
+                n3.append(int(r[:2]))
+            elif q >= -2:
+                n3.append(int(r[:1]))
+
+    r1 = r
+
+    #print n3 # test
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Remove commented python command, leave space after "#" for other comments

File: src/imio/helpers/tests/test_date.py [hunk 5]
```diff
@@ -0,0 +1,30 @@
+# -*- coding: utf-8 -*-
+from imio.helpers.date import formatDate
+from imio.helpers.date import int2word
+from imio.helpers.date import wordizeDate
+from imio.helpers.testing import IntegrationTestCase
+from DateTime import DateTime
```
Comment (gbastien) [inline src/imio/helpers/tests/test_date.py]: Order imports alphabetically

File: src/imio/helpers/date.py [hunk 6]
```diff
@@ -0,0 +1,164 @@
...
+    return converted_date.encode('utf-8') + converted_time.encode('utf-8')
+
+def int2word(n):
+    """
+      Convert an integer number into a string of french word
+      Based on the http://www.daniweb.com/code/snippet609.html
+      code that convert into english words...
+    """
+    ############# globals ################
+    ones = ["", "un ","deux ","trois ","quatre ", "cinq ",
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Please replace constants using uppercased variables (ONES, TENS, ...) and adapt the code

File: src/imio/helpers/date.py [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+            converted_hour = converted_hour[:-1] + "e"
+        converted_minutes = int2word(int(minutes))
+        if hour in ['01', '1', '0', '00', ]:
+            hour_word = "heure"
+        else:
+            hour_word = "heures"
+        #do not display 'zero' in the date like : eleven hour zero...
+        #see http://dev.communesplone.org/trac/ticket/1553
+        if converted_minutes and not minutes in ['0', '00',]:
+            converted_time = " %s %s %s %s" % (u"à", converted_hour, hour_word, converted_minutes)
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Use .format() to be python3 compatible

File: src/imio/helpers/date.py [hunk 8]
```diff
@@ -0,0 +1,164 @@
+# -*- coding: utf-8 -*-
+from DateTime import DateTime
+from imio.helpers.testing import IntegrationTestCase
+from zope.i18n import translate
+
+
+monthsIds = {1:  'month_jan', 2:  'month_feb', 3:  'month_mar', 4:  'month_apr',
+             5:  'month_may', 6:  'month_jun', 7:  'month_jul', 8:  'month_aug',
```
Comment (gbastien) [inline src/imio/helpers/date.py]: CONSTANT too (see comment below)

File: src/imio/helpers/date.py [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+    "quinze ","seize ","dix-sept ","dix-huit ","dix-neuf "]
+    twenties = ["","","vingt ","trente ","quarante ",
+    "cinquante ","soixante ","septante ","quatre-vingt ","nonante "]
+    thousands = ["","mille ","million ", "milliard ", "billiard ",
+    "trilliard ", "quadrilliard ", "quintillard ", "sextilliard", "septilliard ",
+    "octilliard ", "nonilliard ", u"dÃ©cillion ", u"dÃ©cilliard ", u"unodÃ©cillion ", u"unodÃ©cilliard ", u"duodÃ©cillion ", u"duodÃ©cilliard "]
+
+    #quickly manage if n is < 10
+    if n < 10:
+        ones = [u'zéro', "un ","deux ","trois ","quatre ", "cinq ", "six ","sept ","huit ","neuf "]
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Space after ',', is a blank after u'zéro' not needed?  What about something like 0h01?

**Comment by DieKatze** [comment]
Hi again @gbastien,

Are these corrections stopping the bleeding from your eyes?

With kindest regards,

DieKatze.

File: src/imio/helpers/date.py [hunk 10]
```diff
@@ -0,0 +1,177 @@
+# -*- coding: utf-8 -*-
+from DateTime import DateTime
+from imio.helpers.testing import IntegrationTestCase
+from zope.i18n import translate
+
+
+MONTHIDS = {1:  'month_jan', 2:  'month_feb', 3:  'month_mar', 4:  'month_apr',
+             5:  'month_may', 6:  'month_jun', 7:  'month_jul', 8:  'month_aug',
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Please align values in dict

File: src/imio/helpers/date.py [hunk 11]
```diff
@@ -0,0 +1,177 @@
...
+
+def formatDate(date, month_name=True, context=None, long_format=False):
+    """
+      Format the given date.  This is usefull for formatting dates in the pod templates
+      long_format : we want the hour with it
+      mont_name : we want the name of the month instead of the number
+      Note that we force european format...
+    """
+    if date == "now":
+        #we want the current date, calculate it
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Do not forget space after '#'

File: src/imio/helpers/date.py [hunk 12]
```diff
@@ -0,0 +1,177 @@
...
+            elif b1>1:
+                nw = TWENTIES[b2][:-1] + "-" + ONES[b1] + t + nw
+            else:
+                nw = TWENTIES[b2] + extra + ONES[b1] + t + nw
+        if b3 > 0:
+            if b3>1:
+                nw = ONES[b3] + "cent " + nw
+            else:
+                nw = "cent " + nw
+    return nw.strip()
```
Comment (gbastien) [inline src/imio/helpers/date.py]: New line at the end of file? :-)

File: src/imio/helpers/date.py [hunk 13]
```diff
@@ -0,0 +1,177 @@
...
+        month = date.month()
+        translated_month = translate(msgid=MONTHIDS[month],
+                                     domain="plonelocales",
+                                     context=context,
+                                     default=MONTHIDS[month],
+                                     target_language='fr')
+        #use mOntH here to avoid strftime translating the month...
+        res = date.strftime('%e mOntH %Y' + hour)
+        #replace '01 ' by '1er '
+        res = res.replace('01 ', '1er ')
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Please manage day, month and year separately so you can easily replace day of '01' or '1' by '1er' without breaking dates like '2001' that could become '201er'...

File: src/imio/helpers/date.py [hunk 14]
```diff
@@ -0,0 +1,176 @@
...
+    # check if we need to return the month name
+    if month_name:
+        month = date.month()
+        translated_month = translate(msgid=MONTHIDS[month],
+                                     domain="plonelocales",
+                                     context=context,
+                                     default=MONTHIDS[month],
+                                     target_language='fr')
+        # replace '01 ' by '1er '
+        if date.day() in [1, 01]:
```
Comment (gbastien) [inline src/imio/helpers/date.py]: Just doing == 1 should be alright...

File: src/imio/helpers/tests/test_date.py [hunk 15]
```diff
@@ -0,0 +1,40 @@
...
+        self.assertEqual(wordizeDate(DateTime(1989, 9, 19)), 'dix-neuf septembre mille neuf cent quatre-vingt-neuf')
+        self.assertEqual(wordizeDate(DateTime(1, 1, 1)), 'premier janvier deux mille un')
+        self.assertEqual(wordizeDate(DateTime(12, 12, 12, 12, 12), long_format=True), 'douze d\xc3\xa9cembre deux mille douze \xc3\xa0 douze heures douze')
+        self.assertEqual(wordizeDate(DateTime(12, 12, 12, 12), long_format=True), 'douze d\xc3\xa9cembre deux mille douze \xc3\xa0 douze heures')
+        self.assertEqual(wordizeDate(DateTime(1, 1, 1, 1), long_format=True), 'premier janvier deux mille un \xc3\xa0 une heure')
+        self.assertEqual(wordizeDate(DateTime(1, 1, 1, 0, 1), long_format=True), 'premier janvier deux mille un \xc3\xa0 z\xc3\xa9ro heure un')
+
+    def test_formatDate(self):
+        self.assertEqual(formatDate("now", month_name=False), DateTime().strftime("%d/%m/%Y"))
+        self.assertEqual(formatDate(DateTime(2,1,1,1,1), month_name=True, long_format=True), u'1er janvier 2002 (01:01)')
```
Comment (gbastien) [inline src/imio/helpers/tests/test_date.py]: pep8 : E231 : missing whitespace after ','

File: src/imio/helpers/tests/test_date.py [hunk 16]
```diff
@@ -0,0 +1,40 @@
...
+
+
+    def test_int2word(self):
+        self.assertEqual(int2word(21), "vingt et un")
+        self.assertEqual(int2word(100), "cent")
+        self.assertEqual(int2word(220), "deux cent vingt")
+        self.assertEqual(int2word(1000100), "un million cent")
+        self.assertEqual(int2word(123456789876543212345678987654321),
+                         "cent vingt-trois octilliard quatre cent cinquante-six "\
+                         "septilliard sept cent quatre-vingt-neuf sextilliardhuit "\
```
Comment (gbastien) [inline src/imio/helpers/tests/test_date.py]: Problem with "sextilliardhuit"

**Comment by gbastien** [comment]
@DieKatze 

could you please add a line to the changelog?

Thank you!

Gauthier

**Comment by DieKatze** [comment]
@gbastien 

It's done !!!

It was a pleasure and a honor.

Thank you too.

Vivian.

**Pull request merged**